# Regularization

In [1]:
import numpy as np
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

### Recall

In [2]:
X1 = np.array([1,2,3,4,5])
X2 = np.array([2,4,6,8,10])
X = np.c_[X1,X2]

y = X1

In [3]:
model = LinearRegression()
model.fit(X,y)

coefficients = model.coef_
coefficients

array([0.2, 0.4])

### Effect of Ridge or LASSO

In [8]:
ridge_model = Ridge(alpha=1)
ridge_model.fit(X,y)

ridge_model.coef_

array([0.19607843, 0.39215686])

### Questions
1. What will happen if alpha is 0? Same as original model
2. Impact on MSE? Lower accuracy
3. How to choose alpha? Up to you
4. What if I want to consolidate LASSO and Ridge?
   - ElasticNet combines L1 and L2 regularizer. 

In [10]:
mean_squared_error(y,ridge_model.predict(X))

np.float64(0.0007689350249904003)

### Exercises
1. Try different alphas, what will happen to coefficients?
2. Try the same for lasso

In [23]:
def test_ridge(alpha):
    ridge_model = Ridge(alpha=alpha)
    ridge_model.fit(X,y)
    print(f'Ridge coefficient (alpha={alpha}): {ridge_model.coef_}, MSE: {mean_squared_error(y,ridge_model.predict(X))}')

def test_lasso(alpha):
    lasso_model = Lasso(alpha=alpha)
    lasso_model.fit(X,y)
    print(f'Lasso coefficient (alpha={alpha}): {lasso_model.coef_}, MSE: {mean_squared_error(y,lasso_model.predict(X))}')

In [24]:
alphas = [1,10,100]
for a in alphas:
    test_ridge(a)
    test_lasso(a)

Ridge coefficient (alpha=1): [0.19607843 0.39215686], MSE: 0.0007689350249904003
Lasso coefficient (alpha=1): [0.    0.375], MSE: 0.125
Ridge coefficient (alpha=10): [0.16666667 0.33333333], MSE: 0.055555555555555636
Lasso coefficient (alpha=10): [0. 0.], MSE: 2.0
Ridge coefficient (alpha=100): [0.06666667 0.13333333], MSE: 0.8888888888888891
Lasso coefficient (alpha=100): [0. 0.], MSE: 2.0


alphas = [1,10,100]

Ridge coefficient (alpha=1): [0.19607843 0.39215686], MSE: 0.0007689350249904003
Lasso coefficient (alpha=1): [0.    0.375], MSE: 0.0007689350249904003

Ridge coefficient (alpha=10): [0.16666667 0.33333333], MSE: 0.055555555555555636
Lasso coefficient (alpha=10): [0. 0.], MSE: 0.0007689350249904003

Ridge coefficient (alpha=100): [0.06666667 0.13333333], MSE: 0.8888888888888891
Lasso coefficient (alpha=100): [0. 0.], MSE: 0.0007689350249904003